client/core/status.go

```diff
@@ -0,0 +1,382 @@
+		return
+	}
+	// We probably missed an audit request.
+	if len(srvData.MakerSwap) == 0 {
+		err = fmt.Errorf("Server is reporting a match with status MakerSwapCast, but didn't include a coin ID for the swap. %s", logID)
+		return
+	}
+	if len(srvData.MakerContract) == 0 {
+		err = fmt.Errorf("Server is reporting a match with status MakerSwapCast, but didn't include the contract data. %s", logID)
+		return
+	}
+
+	err = trade.auditContract(match, srvData.MakerSwap, srvData.MakerContract)
+	if err != nil {
+		err = fmt.Errorf("auditContract error during match status resolution. %s: %v", logID, err)
+	}
+}
+
+// resolveMissedTakerAudit is a matchConflictResolver to handle the case when
+// our status is MakerSwapCast, but the server is at TakerSwapCast.
+func resolveMissedTakerAudit(c *Core, dc *dexConnection, trade *trackedTrade, match *matchTracker, srvData *msgjson.MatchStatusResult) {
+	logID := statusResolutionID(dc, trade, match)
+	var err error
+	defer func() {
+		if err != nil {
+			match.MetaData.Proof.SelfRevoked = true
+			log.Error(err)
+		}
+	}()
+	// This is nonsensical if we're the taker.
+	if match.Match.Side == order.Taker {
+		err = fmt.Errorf("Server is reporting match in TakerSwapCast, but we're the taker and haven't sent a swap. %s", logID)
+		return
+	}
+	// We probably missed an audit request.
+	if len(srvData.TakerSwap) == 0 {
+		err = fmt.Errorf("Server is reporting a match with status TakerSwapCast, but didn't include a coin ID for the swap. %s", logID)
+		return
+	}
+	if len(srvData.TakerContract) == 0 {
+		err = fmt.Errorf("Server is reporting a match with status TakerSwapCast, but didn't include the contract data. %s", logID)
+		return
+	}
+	err = trade.auditContract(match, srvData.TakerSwap, srvData.TakerContract)
+	if err != nil {
+		err = fmt.Errorf("auditContract error during match status resolution. %s: %v", logID, err)
+	}
+}
+
+// resolveServerMissedMakerInit is a matchConflictResolver to handle the case
+// when our status is MakerSwapCast, but the server is at NewlyMatched.
+func resolveServerMissedMakerInit(c *Core, dc *dexConnection, trade *trackedTrade, match *matchTracker, srvData *msgjson.MatchStatusResult) {
+	logID := statusResolutionID(dc, trade, match)
+	// If we're not the maker, there's nothing we can do.
+	if match.Match.Side != order.Maker {
+		log.Errorf("Server reporting no maker swap, but they've already sent us the swap info. self-revoking. %s", logID)
+		match.MetaData.Proof.SelfRevoked = true
+		return
+
+	}
+	// If we don't have a server acknowledgment, that case will be picked up in
+	// resendPendingRequests at the next tick.
+	if len(match.MetaData.Proof.Auth.InitSig) == 0 {
+		return
+	}
+	// On the other hand, if we do have an acknowledgement from the server,
+	// this appears to be a server error, and we should just revoke the match
+	// and wait to refund.
+	log.Errorf("Server appears to have lost our (maker's) init data after acknowledgement. self-revoking order. %s", logID)
```

Comment:
> Or could `resendPendingRequests` have run concurrently with the resolvers?

It can't, since it the `trackedTrade` mutex is locked for the `tick`, and I've locked the mutexes before requesting `match_status` and they remain locked until after the resolver has run. If a `resendPendingRequests` snuck in before then, the resolver will never actually run because the server would have already been updated, and I check for that condition at https://github.com/decred/dcrdex/blob/5cd4dcc08c61684082cdccf0c44ea7c40f3f8426/client/core/status.go#L133.